tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c

```diff
@@ -0,0 +1,1467 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license.
+// See LICENSE file in the project root for full license information.
+
+#include <stddef.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define ENABLE_MOCKS
+#include "az_pal_os_api.h"
+#include "az_ulib_action_api.h"
+#include "az_ulib_descriptor_api.h"
+#undef ENABLE_MOCKS
+
+#include "az_ulib_ipc_api.h"
+#include "azure_macro_utils/macro_utils.h"
+#include "testrunnerswitcher.h"
+#include "ulib_result.h"
+#include "umock_c/umock_c.h"
+#include "umock_c/umock_c_negative_tests.h"
+#include "umock_c/umocktypes_bool.h"
+#include "umock_c/umocktypes_charptr.h"
+#include "umock_c/umocktypes_stdint.h"
+
+static TEST_MUTEX_HANDLE g_test_by_test;
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code) {
+  ASSERT_FAIL("umock_c reported error :%s", MU_ENUM_TO_STRING(UMOCK_C_ERROR_CODE, error_code));
+}
+
+static uint32_t my_property = 0;
+
+static AZ_ULIB_RESULT get_my_property(const void *model_out) {
+  uint32_t *new_val = (uint32_t *)model_out;
+
+  *new_val = my_property;
+
+  return AZ_ULIB_SUCCESS;
+}
+
+static AZ_ULIB_RESULT set_my_property(const void *const model_in) {
+  uint32_t *new_val = (uint32_t *)model_in;
+
+  my_property = *new_val;
+
+  return AZ_ULIB_SUCCESS;
+}
+
+typedef struct my_method_model_in_tag {
+  uint8_t action;
+  const az_ulib_interface_descriptor *descriptor;
+  az_ulib_ipc_interface_handle handle;
+  AZ_ULIB_RESULT return_result;
+} my_method_model_in;
+
+static AZ_ULIB_RESULT my_method(const void *const model_in, const void *model_out) {
+  my_method_model_in *in = (my_method_model_in *)model_in;
+  AZ_ULIB_RESULT *result = (AZ_ULIB_RESULT *)model_out;
+
+  switch (in->action) {
+    case 0:
+      *result = in->return_result;
+      break;
+    case 1:
+      *result = az_ulib_ipc_unpublish(in->descriptor, AZ_ULIB_NO_WAIT);
+      break;
+    case 2:
+      *result = az_ulib_ipc_release_interface(in->handle);
+      break;
+    default:
+      *result = AZ_ULIB_NO_SUCH_ELEMENT_ERROR;
+      break;
+  }
+
+  return AZ_ULIB_SUCCESS;
+}
+
+static AZ_ULIB_RESULT my_method_async(
+    const void *const model_in,
+    const void *model_out,
+    const az_ulib_action_token action_token,
+    az_ulib_action_cancelation_callback *cancel) {
+  (void)model_in;
+  (void)model_out;
+  (void)action_token;
+  (void)cancel;
+
+  return AZ_ULIB_SUCCESS;
+}
+
+static AZ_ULIB_RESULT my_method_cancel(const az_ulib_action_token action_token) {
+  (void)action_token;
+
+  return AZ_ULIB_SUCCESS;
+}
+
+#define MY_METHOD_HANDLE 3
+
+AZ_ULIB_DESCRIPTOR_CREATE(
+    MY_INTERFACE_1_123,
```

Comment:
>MY_INTERFACE_1_123 [](start = 4, length = 18)

suggestion don't have to. but for each of these it might be clearer to name it `MY_INTERFACE_1_V123` to specify the last number is the version. makes it a little easier to parse through. i was reading it and took a little to link the different numbers with their purpose.